### PR TITLE
feat(graphql): rewrite GraphQL data provider

### LINKED ISSUE #5942: [BUG] Default GraphQL DataProvider GetListResponse from 'count' property not possible.
### Describe the bug

The default GraphQL providers `getList` function return the following `GetListResponse` shaped object: 

https://github.com/refinedev/refine/blob/3046f9610f647bb049dbd388cd8cd5c267b481fa/packages/graphql/src/dataProvider/index.ts#L39-L42

As [mentioned by another user on Discord](https://discord-questions.refine.dev/m/1205328662510047242) (possibly @Karabur?), this does not make sense. In order for `getList` to function correctly with various integration points within Refine (like the MUI `useDataGrid` hook), `response[operation]` must return an array of objects. An array cannot have additional properties on it such as `count`.  I'm not sure I understand how anybody could be using this default data provider out of the box with working pagination. 

### Steps To Reproduce

N/A. Per the graphQL spec, a list cannot have additional properties. 

### Expected behavior

I do not see how any GraphQL API could provide a response that would provide working pagination for the default GraphQL provider. Even the [docs for useList indicate something considerably different](https://refine.dev/docs/data/hooks/use-list/#retrieving-the-total-row-count), more akin to a [Relay response](https://relay.dev/graphql/connections.htm).

### Packages

- @refinedev/graphql: 6.5.0

### Additional Context

_No response_

### LINKED ISSUE #5943: [FEAT] Default GraphQL DataProvider cannot function on deeper graphs
### Is your feature request related to a problem? Please describe.

Currently, the [default GraphQL DataProvider](https://github.com/refinedev/refine/blob/3046f9610f647bb049dbd388cd8cd5c267b481fa/packages/graphql/src/dataProvider/index.ts) is not capable of handling complex object graphs (which is very common in GraphQL). It appears to make the assumption that any query will be designed such that it returns the intended data type at the root level of the response. But this is very frequently not the case with GraphQL. 

E.g. 

```graphql
query {
  userPosts(userId: 'me') {
    id
    title
    content
  }
}
```

But as is typically the case with GraphQL endpoints, they may not expose highly focused queries such as that, and you may have something along the lines of this, which is still very simplistic but contains the desired data only a single level deeper.

```graphql
query {
  user(userId: 'me') {
    posts {
      id
      title
      content
    }
  }
}
```

Unfortunately this is not possible with the current DataProvider. 

### Describe alternatives you've considered

I propose amending the [GraphQLQueryOptions](https://github.com/refinedev/refine/blob/3046f9610f647bb049dbd388cd8cd5c267b481fa/packages/core/src/contexts/data/types.ts#L49-L99) type with a `gqlDataPath` and a `gqlCountPath` properties. These would be an array of strings that tell the DataProvider how to reach the intended data from the response. 

An additional utils function could be added that would then take this path (if provided) and extract the intended object(s) for use in the response, e.g...

```typescript
const getProp = (obj: any, path: (string | number)[], defaultValue: any = undefined): any =>
  path.reduce((acc: any, key: string | number) => acc && acc[key] !== undefined ? acc[key] : defaultValue, obj);
```

Then the returned result could be derived as so...

```typescript
return {
  data: meta?.gqlDataPath ? getProp(meta.gqlDataPath) : response[operation],
  total: meta?.gqlCountPath ? getProp(meta.gqlCountPath): response[operation].count
}
```

### Additional context

_No response_

### Describe the thing to improve

This gives the default GraphQL DataProvider the ability to handle more than the most basic of response types. In my experience, one of the primary purposes of graphQL is to expose deeper, more complex graphs of data precisely because the data owners may not be aware of how clients intend to use it. It's likely not possible to expose rudimentary endpoints for every possible use case. GitHub's GraphQL endpoint is an excellent example of this.

### PR DESCRIPTION
**Pull request by BatuhanW**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://refine.dev/docs/guides-concepts/contributing/#commit-convention

## Bugs / Features

- [x] Related issue(s) linked
- [x] Tests for the changes have been added
- [x] Docs have been added / updated
- [x] Changesets have been added https://refine.dev/docs/guides-concepts/contributing/#creating-a-changeset

## What is the new behavior?

fixes #5942 #5943

### CODE REVIEW

**Review by omeraplak — APPROVED** [review]

File: packages/graphql/src/dataProvider/options.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+};
+
+type GraphQLGetCountFunction = (
+  params: GraphQLGetCountFunctionParams,
+) => number;
+
+export type GraphQLDataProviderOptions = {
+  getData: GraphQLGetDataFunction;
+  getCount: GraphQLGetCountFunction;
+};
```
Comment (aliemir) [inline packages/graphql/src/dataProvider/options.ts:36]: This type is also used in the `dataProvider` export and forces user to define both of the functions. It should allow partial assignments.

**Review by aliemir — CHANGES_REQUESTED** [review]
Left couple of comments for types, typos and conventions. Thank you 🙏

File: packages/graphql/src/dataProvider/newOptions.ts
Comment (aliemir) [inline packages/graphql/src/dataProvider/newOptions.ts:1]: Unwanted file here.

File: packages/graphql/src/utils/buildFilters.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
+import type { CrudOperators, LogicalFilter, CrudFilter } from "@refinedev/core";
+import set from "lodash/set";
```
Comment (aliemir) [inline packages/graphql/src/utils/buildFilters.ts:2]: `lodash` dependency is missing from `package.json`

**Review by aliemir — CHANGES_REQUESTED** [review]

File: packages/graphql/src/dataProvider/options.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+      response: OperationResult<any>,
+      params: UpdateManyParams<any>,
+    ) => {
+      const pascalResource = camelcase(params.resource, {
+        pascalCase: true,
+      });
+
+      const key = `updateMany${pascalResource}`;
+      return response.data?.[key];
+    },
```
[the remark this answers is not in the document]
Comment (BatuhanW) [reply packages/graphql/src/dataProvider/options.ts:156]: It's handled here: https://github.com/refinedev/refine/blob/936e9fc80ab3b94cc4dfec0bda8e41442011f747/packages/graphql/src/dataProvider/index.ts#L138

[the remark this answers is not in the document]
Comment (aliemir) [reply packages/graphql/src/dataProvider/options.ts]: Not really. This returns a single item in an array. `updateMany` should return as many items in an array as the provided number of ids.

[the remark this answers is not in the document]
Comment (BatuhanW) [reply packages/graphql/src/dataProvider/options.ts]: We shouldn't force it. For example these are valid responses:

```
mutation UpdateManyBlogPosts($input: UpdateManyBlogPostsInput!) {
  updateManyBlogPosts(input: $input) {
    updatedCount
  }
}

mutation DeleteManyBlogPosts($input: DeleteManyBlogPostsInput!) {
  deleteManyBlogPosts(input: $input) {
    deletedCount
  }
}
```

**Review by aliemir — APPROVED** [review]

**Pull request merged**